node/actors/direct_start_agent.go

```diff
@@ -32,17 +44,111 @@ func (a *DirectStartAgent) PostStop(ctx context.Context) error {
 	case *goaktpb.PostStart:
-		a.logger = ctx.Self().Logger()
-		a.logger.Infof("Direct start agent '%v' is running", ctx.Self().Name())
+		a.self = ctx.Self()
+		a.logger.Info("Direct start agent is running", slog.String("name", ctx.Self().Name()))
+	case *actorproto.StartWorkload:
+		a.logger.Debug("StartWorkload received", slog.String("name", ctx.Self().Name()), slog.String("workload", m.WorkloadName))
+		resp, err := a.startWorkload(m)
+		if err != nil {
+			ctx.Err(err)
+			return
+		}
+		ctx.Response(resp)
+	case *actorproto.StopWorkload:
+		a.logger.Debug("StopWorkload received", slog.String("name", ctx.Self().Name()), slog.String("workload", m.WorkloadId))
+		resp, err := a.stopWorkload(m)
+		if err != nil {
+			ctx.Err(err)
+			return
+		}
+		ctx.Response(resp)
 	case *actorproto.QueryWorkloads:
-		a.queryWorkloads(ctx)
+		a.logger.Debug("QueryWorkloads received", slog.String("name", ctx.Self().Name()))
+		resp, err := a.queryWorkloads(m)
+		if err != nil {
+			ctx.Err(err)
+			return
+		}
+		ctx.Response(resp)
+	case *goaktpb.Terminated:
 	default:
+		a.logger.Warn("Direct start agent received unhandled message", slog.String("name", ctx.Self().Name()), slog.Any("message_type", fmt.Sprintf("%T", m)))
 		ctx.Unhandled()
 	}
 }
 
-func (a *DirectStartAgent) queryWorkloads(ctx *goakt.ReceiveContext) {
-	// TODO: make this real
+func (a *DirectStartAgent) startWorkload(m *actorproto.StartWorkload) (*actorproto.WorkloadStarted, error) {
+	env := make(map[string]string)
+	if m.Environment != "" {
+		env_b, err := base64.StdEncoding.DecodeString(m.Environment)
+		if err != nil {
+			a.logger.Error("Failed to decode environment", slog.String("name", a.self.Name()), slog.Any("err", err))
+			return nil, err
+		}
+
+		err = json.Unmarshal(env_b, &env)
+		if err != nil {
+			a.logger.Error("Failed to unmarshal environment", slog.String("name", a.self.Name()), slog.Any("err", err))
+			return nil, err
+		}
+	}
+
+	ref, err := getArtifact(m.WorkloadName, m.Uri)
+	if err != nil {
+		a.logger.Error("Failed to get artifact", slog.String("name", a.self.Name()), slog.Any("err", err))
+		return nil, err
+	}
+
+	workloadId := nuid.New().Next()
+	pa, err := createNewProcessActor(a.logger.WithGroup(workloadId), workloadId, m, ref, env)
+	if err != nil {
+		a.logger.Error("Failed to create process actor", slog.String("name", a.self.Name()), slog.Any("err", err))
+		return nil, err
+	}
+
+	c, err := a.self.SpawnChild(context.Background(), workloadId, pa)
+	if err != nil {
+		a.logger.Error("Failed to spawn child", slog.String("name", a.self.Name()), slog.String("workload", m.WorkloadName), slog.Any("err", err))
+		return nil, err
+	}
+
+	a.logger.Info("Spawned direct start process", slog.String("name", c.Name()))
+	return &actorproto.WorkloadStarted{Id: workloadId, Name: m.WorkloadName, Started: true}, nil
+}
+
+func (a *DirectStartAgent) stopWorkload(m *actorproto.StopWorkload) (*actorproto.WorkloadStopped, error) {
+	for _, c := range a.self.Children() {
```

Comment:
you don't need to range through all children here. Instead use `Child(workloadId)` to get the specific one that you want.

Reply:
did copilot give you this?